{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Imports"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib as mpl\n",
    "import seaborn as sns\n",
    "import random\n",
    "import scipy.stats\n",
    "\n",
    "from pathlib import Path\n",
    "from Bio import Phylo\n",
    "from sklearn.metrics import roc_auc_score\n",
    "from sklearn.metrics import roc_curve\n",
    "from sklearn.metrics import RocCurveDisplay\n",
    "\n",
    "\n",
    "import warnings\n",
    "warnings.filterwarnings('ignore')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "Data_folder = Path(\"data/local\")\n",
    "Figures_folder = Path(\"Figures\")\n",
    "Figures_folder.mkdir(parents=True, exist_ok=True)\n",
    "\n",
    "\n",
    "stats_df_path = Data_folder / \"stats_ESC_GA4805AA.csv\"\n",
    "aa_site_df_path = Data_folder / \"mutants_sites_ESC_GA4805AA.csv\"\n",
    "scores_df_path = Data_folder / \"scores_subset.csv\"\n",
    "full_seq_df_path = Data_folder / \"full_seq_single_muts.csv\"\n",
    "double_muts_df_path = Data_folder / \"double_mut_epistasis.csv\"\n",
    "ipr_df_path = Data_folder / \"IPR.csv\"\n",
    "couplings_df_path = Data_folder / \"couplings.csv\"\n",
    "simulated_sites_df_path = Data_folder / \"simulated_sites_ESC_GA4805AA.csv\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Figure parameters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "linewidth = 1\n",
    "SMALL_SIZE = 8\n",
    "MEDIUM_SIZE = 12\n",
    "BIGGER_SIZE = 16\n",
    "DPI = 300\n",
    "WIDTH = 3.5\n",
    "HW_RATIO = 2/3\n",
    "\n",
    "def parameters():\n",
    "    plt.rc('font', size=SMALL_SIZE, family='sans-serif') \n",
    "    plt.rc('axes', titlesize=SMALL_SIZE)     # fontsize of the axes title\n",
    "    plt.rc('axes', labelsize=MEDIUM_SIZE)    # fontsize of the x and y labels\n",
    "    plt.rc('xtick', labelsize=SMALL_SIZE)    # fontsize of the tick labels\n",
    "    plt.rc('ytick', labelsize=SMALL_SIZE)    # fontsize of the tick labels\n",
    "    plt.rc('legend', fontsize=SMALL_SIZE)    # legend fontsize\n",
    "    plt.rc('legend', title_fontsize=MEDIUM_SIZE)    # legend fontsize\n",
    "    plt.rc('figure', titlesize=BIGGER_SIZE)  # fontsize of the figure title\n",
    "\n",
    "color_variable = sns.color_palette(\"colorblind\")[0]\n",
    "color_conserved = sns.color_palette(\"Set2\")[4]\n",
    "color_alw_conserved = sns.color_palette(\"Paired\")[2]\n",
    "color_cont_conserved = sns.color_palette(\"Paired\")[3]\n",
    "color_dca = sns.color_palette(\"Paired\")[5]\n",
    "color_ind = sns.color_palette(\"Set2\")[5]\n",
    "color_rand = sns.color_palette(\"tab10\")[7]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Analysis parameters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "prop_gaps = 0.2"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Clean dataframes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "stats_df = pd.read_csv(stats_df_path)\n",
    "stats_df = stats_df[(stats_df[\"Prop_gaps_local\"]<prop_gaps) & (stats_df[\"Prop_gaps_distant\"]<prop_gaps)]\n",
    "stats_df[\"IG\"] = stats_df[\"CI_Entropy\"]-stats_df[\"CD_Entropy\"]\n",
    "\n",
    "aa_site_df = pd.read_csv(aa_site_df_path)\n",
    "aa_site_df = stats_df[[\"Gene\",\"Locus\"]].merge(aa_site_df,how=\"left\")\n",
    "aa_site_df = aa_site_df.sort_values(by=[\"Gene\",\"Locus\",\"DCA_score\"])\n",
    "aa_site_df[\"Rank_DCA\"] = [i for i in range(20)]*len(aa_site_df[[\"Gene\",\"Locus\"]].drop_duplicates())\n",
    "aa_site_df = aa_site_df.sort_values(by=[\"Gene\",\"Locus\",\"Profile\"], ascending=[True,True,False])\n",
    "aa_site_df[\"Rank_Ind\"] = [i for i in range(20)]*len(aa_site_df[[\"Gene\",\"Locus\"]].drop_duplicates())\n",
    "aa_site_df = stats_df[[\"Gene\",\"Locus\",\"CD_Entropy\"]].merge(aa_site_df)\n",
    "aa_site_df[\"exp_DCA\"] = np.exp(-aa_site_df[\"DCA_score\"])\n",
    "norm = aa_site_df[[\"Gene\",\"Locus\",\"exp_DCA\"]].groupby([\"Gene\",\"Locus\"]).sum().reset_index()\n",
    "norm.columns = [\"Gene\",\"Locus\",\"Norm\"]\n",
    "aa_site_df = aa_site_df.merge(norm, on=[\"Gene\",\"Locus\"])\n",
    "aa_site_df[\"DCA_Proba\"] = aa_site_df[\"exp_DCA\"]/aa_site_df[\"Norm\"]\n",
    "columns = ['Gene', 'Locus', 'CD_Entropy', 'Reference', 'AA', 'Observed', 'Profile','Accessible', 'DCA_score', 'Rank_DCA', 'Rank_Ind', 'DCA_Proba']\n",
    "aa_site_df = aa_site_df[columns]\n",
    "aa_site_df[\"acc_DCA_Proba\"] = aa_site_df[\"DCA_Proba\"]*(aa_site_df[\"Accessible\"]).astype(int)\n",
    "norm = aa_site_df[[\"Gene\",\"Locus\",\"acc_DCA_Proba\"]].groupby([\"Gene\",\"Locus\"]).sum().reset_index()\n",
    "norm.columns = [\"Gene\",\"Locus\",\"Norm\"]\n",
    "aa_site_df = aa_site_df.merge(norm, on=[\"Gene\",\"Locus\"])\n",
    "aa_site_df[\"acc_DCA_Proba\"] = aa_site_df[\"acc_DCA_Proba\"]/aa_site_df[\"Norm\"]\n",
    "columns += [\"acc_DCA_Proba\"]\n",
    "aa_site_df = aa_site_df[columns]\n",
    "aa_site_df[\"acc_Profile\"] = aa_site_df[\"Profile\"]*(aa_site_df[\"Accessible\"]).astype(int)\n",
    "norm = aa_site_df[[\"Gene\",\"Locus\",\"acc_Profile\"]].groupby([\"Gene\",\"Locus\"]).sum().reset_index()\n",
    "norm.columns = [\"Gene\",\"Locus\",\"Norm\"]\n",
    "aa_site_df = aa_site_df.merge(norm, on=[\"Gene\",\"Locus\"])\n",
    "aa_site_df[\"acc_Profile\"] = aa_site_df[\"acc_Profile\"]/aa_site_df[\"Norm\"]\n",
    "columns += [\"acc_Profile\"]\n",
    "aa_site_df = aa_site_df[columns]\n",
    "\n",
    "couplings_df = pd.read_csv(couplings_df_path)\n",
    "couplings_df = stats_df[[\"Gene\",\"Locus\"]].merge(couplings_df, right_on=[\"Gene\",\"i\"], left_on=[\"Gene\", \"Locus\"],how=\"inner\").merge(stats_df[[\"Gene\",\"Locus\"]], left_on=[\"Gene\",\"k\"], right_on=[\"Gene\", \"Locus\"],how=\"inner\")\n",
    "couplings_df = couplings_df[(couplings_df[\"AAi1\"]!=\"-\") & (couplings_df[\"AAi2\"]!=\"-\") & (couplings_df[\"AAk1\"]!=\"-\") & (couplings_df[\"AAk2\"]!=\"-\")]\n",
    "couplings_df[\"Coupling\"] = couplings_df[\"J_i1k1\"]+couplings_df[\"J_i2k2\"]-(couplings_df[\"J_i1k2\"]+couplings_df[\"J_i2k1\"])\n",
    "\n",
    "full_seq_df = pd.read_csv(full_seq_df_path)\n",
    "full_seq_df[\"Diff\"] = full_seq_df[\"Full_seq\"]-full_seq_df[\"Single_muts\"]\n",
    "full_seq_df[\"Cat\"] = pd.cut(full_seq_df[\"FD\"],bins=[0,0.05,0.1,0.15,0.2,0.25,1.1], labels=[\"<5%\",\"5-10%\",\"10-15%\",\"15-20%\",\"20-25%\", \">25%\"])\n",
    "\n",
    "double_muts = pd.read_csv(double_muts_df_path)\n",
    "col = double_muts.columns\n",
    "double_muts = stats_df[[\"Gene\",\"Locus\"]].merge(double_muts, left_on=[\"Gene\",\"Locus\"], right_on=[\"Gene\",\"Loci_1\"], how=\"inner\")[col].merge(stats_df, left_on=[\"Gene\",\"Loci_2\"], right_on=[\"Gene\",\"Locus\"], how=\"inner\")[col]\n",
    "\n",
    "ipr_df = pd.read_csv(ipr_df_path)\n",
    "size = ipr_df[[\"Gene\",\"Locus\"]].groupby(\"Gene\").max().reset_index()\n",
    "size.columns = [\"Gene\", \"Length\"]\n",
    "ipr_df = stats_df[[\"Gene\",\"Locus\",\"CI_Entropy\",\"CD_Entropy\"]].merge(ipr_df).merge(size)\n",
    "ipr_df[\"Neff\"] = 1/(ipr_df[\"IPR\"]*ipr_df[\"Length\"])\n",
    "ipr_df[\"IG\"] = ipr_df[\"CI_Entropy\"]-ipr_df[\"CD_Entropy\"]\n",
    "\n",
    "print(\"{} sites in {} domains across {} genes\".format(len(stats_df),len(np.unique(stats_df[\"Gene\"])), len(np.unique(stats_df[\"Gene\"].apply(lambda x: x.split(\"-\")[1])))))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Figures\n",
    "## Subsection 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "aa_site_df_1 = aa_site_df[['Gene', 'Locus', 'CD_Entropy', 'Reference', 'AA', 'Observed', 'Profile', 'Accessible', 'DCA_score', 'Rank_DCA']]\n",
    "aa_site_df_1.columns = ['Gene', 'Locus', 'CD_Entropy', 'Reference', 'AA', 'Observed', 'Profile', 'Accessible', 'DCA_score', 'Rank']\n",
    "aa_site_df_1[\"Model\"] = \"DCA model\"\n",
    "aa_site_df_2 = aa_site_df[['Gene', 'Locus', 'CD_Entropy', 'Reference', 'AA', 'Observed', 'Profile', 'Accessible', 'DCA_score', 'Rank_Ind']]\n",
    "aa_site_df_2.columns = ['Gene', 'Locus', 'CD_Entropy', 'Reference', 'AA', 'Observed', 'Profile', 'Accessible', 'DCA_score', 'Rank']\n",
    "aa_site_df_2[\"Model\"] = \"IND model\"\n",
    "aa_site_df_1 = pd.concat([aa_site_df_1,aa_site_df_2])\n",
    "aa_site_df_2 = None\n",
    "aa_site_df_1 = aa_site_df_1[aa_site_df_1[\"Reference\"]]\n",
    "tot = len(aa_site_df_1[aa_site_df_1[\"Model\"]==\"DCA model\"])\n",
    "aa_site_df_1 = aa_site_df_1.groupby([\"Rank\",\"Model\"]).size().reset_index()\n",
    "aa_site_df_1.columns = [\"Rank\",\"Model\",\"Nb\"]\n",
    "aa_site_df_1[\"Prop\"] = aa_site_df_1[\"Nb\"]/tot\n",
    "aa_site_df_1[\"Rank\"] = aa_site_df_1[\"Rank\"]+1\n",
    "\n",
    "with sns.axes_style(\"whitegrid\"):\n",
    "    parameters()  \n",
    "    fig = sns.catplot(x=\"Rank\", y=\"Prop\", hue=\"Model\", kind=\"bar\", data=aa_site_df_1, height=WIDTH*HW_RATIO, palette=[color_dca,color_ind], alpha=1, aspect=1/HW_RATIO, legend_out=False)\n",
    "    plt.legend(loc='upper right')\n",
    "    fig.set_axis_labels(\"Amino Acid Rank\",\"Proportion\")\n",
    "    fig.savefig(Figures_folder / \"Fig_1a.jpeg\", dpi=DPI, bbox_inches=\"tight\")\n",
    "\n",
    "print(f\"1st rank DCA: {aa_site_df_1[(aa_site_df_1['Rank']==1) & (aa_site_df_1['Model']=='DCA model')].loc[0,'Prop']}\")\n",
    "print(f\"1st rank IND: {aa_site_df_1[(aa_site_df_1['Rank']==1) & (aa_site_df_1['Model']=='IND model')].loc[1,'Prop']}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "aa_site_df_1 = aa_site_df[['Gene', 'Locus', 'CD_Entropy', 'Reference', 'AA', 'Observed', 'Profile', 'Accessible', 'DCA_score', 'Rank_DCA']]\n",
    "aa_site_df_1.columns = ['Gene', 'Locus', 'CD_Entropy', 'Reference', 'AA', 'Observed', 'Profile', 'Accessible', 'DCA_score', 'Rank']\n",
    "aa_site_df_1[\"Model\"] = \"DCA\"\n",
    "aa_site_df_2 = aa_site_df[['Gene', 'Locus', 'CD_Entropy', 'Reference', 'AA', 'Observed', 'Profile', 'Accessible', 'DCA_score', 'Rank_Ind']]\n",
    "aa_site_df_2.columns = ['Gene', 'Locus', 'CD_Entropy', 'Reference', 'AA', 'Observed', 'Profile', 'Accessible', 'DCA_score', 'Rank']\n",
    "aa_site_df_2[\"Model\"] = \"Independent\"\n",
    "aa_site_df_1 = pd.concat([aa_site_df_1,aa_site_df_2])\n",
    "aa_site_df_2 = None\n",
    "aa_site_df_1 = aa_site_df_1[(~aa_site_df_1[\"Reference\"]) & (aa_site_df_1[\"Observed\"]>0.05)]\n",
    "tot = len(aa_site_df_1[aa_site_df_1[\"Model\"]==\"DCA\"])\n",
    "aa_site_df_1 = aa_site_df_1.groupby([\"Rank\",\"Model\"]).size().reset_index()\n",
    "aa_site_df_1.columns = [\"Rank\",\"Model\",\"Nb\"]\n",
    "aa_site_df_1[\"Prop\"] = aa_site_df_1[\"Nb\"]/tot\n",
    "aa_site_df_1[\"Rank\"] = aa_site_df_1[\"Rank\"]+1\n",
    "\n",
    "polymorphism = aa_site_df[(aa_site_df[\"Observed\"]>0.05) & (aa_site_df[\"Observed\"]<=0.5)][[\"Gene\",\"Locus\",\"Rank_DCA\"]] # & (aa_site_df[\"CD_Entropy\"]<3)\n",
    "consensus = aa_site_df[aa_site_df[\"Observed\"]>0.5][[\"Gene\",\"Locus\",\"Rank_DCA\"]]\n",
    "consensus = polymorphism[[\"Gene\",\"Locus\"]].merge(consensus,how=\"inner\")\n",
    "polymorphism[\"Cat\"] = \"Minor allele\"\n",
    "polymorphism[\"Total\"] = len(polymorphism)\n",
    "consensus[\"Cat\"] = \"Major allele\"\n",
    "consensus[\"Total\"] = len(consensus)\n",
    "polymorphism = pd.concat([polymorphism, consensus])\n",
    "consensus = None\n",
    "polymorphism = polymorphism.groupby([\"Rank_DCA\",\"Cat\",\"Total\"]).size().reset_index()\n",
    "polymorphism.columns = [\"Rank_DCA\",\"Allele\",\"Total\",\"Nb\"]\n",
    "polymorphism[\"Prop\"] = polymorphism[\"Nb\"]/polymorphism[\"Total\"]\n",
    "polymorphism[\"Rank_DCA\"] = polymorphism[\"Rank_DCA\"]+1\n",
    "polymorphism.head()\n",
    "with sns.axes_style(\"whitegrid\"):\n",
    "    parameters() \n",
    "    fig = sns.catplot(x=\"Rank_DCA\", y=\"Prop\", hue=\"Allele\", kind=\"bar\", data=polymorphism, height=WIDTH*HW_RATIO,\n",
    "                      palette=[sns.color_palette(\"Paired\")[5],sns.color_palette(\"Paired\")[4]], alpha=1, aspect=1/HW_RATIO, legend_out=False)\n",
    "    fig.set_axis_labels(\"Amino Acid Rank (DCA model)\",\"Proportion\")\n",
    "    plt.legend(loc='upper right')\n",
    "    fig.savefig(Figures_folder / \"Fig_1b.jpeg\", dpi=DPI, bbox_inches=\"tight\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "polymorphism.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "th_poly = 0.05\n",
    "epsilon = 0.01\n",
    "\n",
    "aa_site_df_1 = aa_site_df[~aa_site_df[\"Reference\"]]\n",
    "aa_site_df_1[\"Proba\"] = aa_site_df_1[\"Profile\"]/aa_site_df_1[\"Profile\"].sum()\n",
    "N = len(aa_site_df[(~aa_site_df[\"Reference\"]) & (aa_site_df[\"Observed\"]>th_poly)])\n",
    "DCA = list(aa_site_df[(~aa_site_df[\"Reference\"]) & (aa_site_df[\"Observed\"]>th_poly)][\"DCA_score\"])\n",
    "DCA += list(aa_site_df_1.sample(n=N, replace=False, weights=\"Proba\")[\"DCA_score\"])\n",
    "DCA += list(aa_site_df[(~aa_site_df[\"Reference\"])].sample(n=N, replace=False)[\"DCA_score\"])\n",
    "Label = [\"Obs\"]*N + [\"Profile\"]*N + [\"Random\"]*N\n",
    "df = pd.DataFrame({\"DCA\":DCA, \"Label\":Label})\n",
    "\n",
    "min_, max_ = aa_site_df[\"DCA_score\"].min()-epsilon, aa_site_df[\"DCA_score\"].max()+epsilon\n",
    "bins = np.linspace(min_,max_,40)\n",
    "\n",
    "with sns.axes_style(\"whitegrid\"):\n",
    "    parameters()   \n",
    "    fig, (ax_box, ax_hist) = plt.subplots(2, figsize=(WIDTH,WIDTH*HW_RATIO*1.7), sharex=True, gridspec_kw={\"height_ratios\": (.15, .85)}, dpi=DPI)\n",
    "    sns.boxplot(x=\"DCA\", y=\"Label\", order=[\"Obs\", \"Profile\", \"Random\"], palette=[color_variable, color_ind, color_rand], data=df, ax=ax_box, linewidth=linewidth, fliersize=linewidth)\n",
    "    sns.histplot(data=df[(df[\"Label\"]==\"Obs\")], x=\"DCA\", color=color_variable, kde=True, label=\"Observed polymorphisms\",ax=ax_hist,bins=bins,line_kws={\"linewidth\":linewidth})\n",
    "    sns.histplot(data=df[(df[\"Label\"]==\"Profile\")], x=\"DCA\", color=color_ind, kde=True, label=\"Profile polymorphisms\",ax=ax_hist,bins=bins,line_kws={\"linewidth\":linewidth})\n",
    "    sns.histplot(data=df[(df[\"Label\"]==\"Random\")], x=\"DCA\", color=color_rand, kde=True, label=\"Random polymorphisms\",ax=ax_hist,bins=bins,line_kws={\"linewidth\":linewidth})\n",
    "    ax_box.set(xlabel='')\n",
    "    ax_box.set(ylabel='')\n",
    "    ax_box.set_yticks([])\n",
    "    ax_hist.set_xlabel(\"DCA score\")\n",
    "    ax_hist.set_ylabel(\"Number of polymorphisms\")\n",
    "    ax_hist.set_ylim((0,500))\n",
    "    plt.legend(loc='upper right')   \n",
    "    fig.savefig(Figures_folder / \"Fig_1c.jpeg\", dpi=DPI, bbox_inches=\"tight\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = aa_site_df[((aa_site_df[\"Observed\"]>th_poly) | (aa_site_df[\"Observed\"]==0)) & (~aa_site_df[\"Reference\"])] \n",
    "df = df[~df[\"Gene\"].isin(['PF00033-GA4805AA_01620', 'PF01306-GA4805AA_02748', 'PF04403-GA4805AA_01031'])]\n",
    "print(roc_auc_score((df[\"Observed\"]>th_poly).astype(int), df[\"acc_Profile\"])-roc_auc_score((df[\"Observed\"]>th_poly).astype(int), df[\"Profile\"]))\n",
    "print(roc_auc_score((df[\"Observed\"]>th_poly).astype(int), df[\"acc_DCA_Proba\"])-roc_auc_score((df[\"Observed\"]>th_poly).astype(int), df[\"DCA_Proba\"]))\n",
    "\n",
    "\n",
    "with sns.axes_style(\"whitegrid\"):\n",
    "    parameters()\n",
    "    fig = plt.figure(figsize=(WIDTH,WIDTH*HW_RATIO), dpi=DPI)\n",
    "    ax = fig.add_subplot(111)\n",
    "\n",
    "    fpr, tpr, threshold = roc_curve((df[\"Observed\"]>th_poly).astype(int), df[\"Profile\"])\n",
    "    display = RocCurveDisplay(fpr=fpr, tpr=tpr)\n",
    "    display.plot(ax=ax, label=\"Independent model:\\n{}\".format(np.round(roc_auc_score((df[\"Observed\"]>th_poly).astype(int), df[\"Profile\"]),3)), color=color_ind, linewidth=linewidth) \n",
    "\n",
    "    fpr, tpr, threshold = roc_curve((df[\"Observed\"]>th_poly).astype(int), df[\"acc_Profile\"])\n",
    "    display = RocCurveDisplay(fpr=fpr, tpr=tpr)\n",
    "    display.plot(ax=ax, label=\"Independent model\\n(1-SNP): {}\".format(np.round(roc_auc_score((df[\"Observed\"]>th_poly).astype(int), df[\"acc_Profile\"]),3)), color=color_ind, linestyle=\"--\", linewidth=linewidth)\n",
    "\n",
    "    fpr, tpr, threshold = roc_curve((df[\"Observed\"]>th_poly).astype(int), df[\"DCA_Proba\"])\n",
    "    display = RocCurveDisplay(fpr=fpr, tpr=tpr)\n",
    "    display.plot(ax=ax, label=\"DCA model: {}\".format(np.round(roc_auc_score((df[\"Observed\"]>th_poly).astype(int), df[\"DCA_Proba\"]),3)), color=color_dca, linewidth=linewidth)\n",
    "\n",
    "    fpr, tpr, threshold = roc_curve((df[\"Observed\"]>th_poly).astype(int), df[\"acc_DCA_Proba\"])\n",
    "    display = RocCurveDisplay(fpr=fpr, tpr=tpr)\n",
    "    display.plot(ax=ax, label=\"DCA model (1-SNP):\\n{}\".format(np.round(roc_auc_score((df[\"Observed\"]>th_poly).astype(int), df[\"acc_DCA_Proba\"]),3)), color=color_dca, linestyle=\"--\", linewidth=linewidth) \n",
    "    fig.savefig(Figures_folder / \"Fig_S2.jpeg\", dpi=DPI, bbox_inches=\"tight\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Subsection 3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig = sns.jointplot(x=\"CD_Entropy\",y=\"CI_Entropy\",data=stats_df,kind=\"hex\",cmap='CMRmap_r',height=WIDTH) #cubehelix_r, gist_earth_r, afmhot_r, hot_r, gnuplot2_r\n",
    "\n",
    "for patch in fig.ax_marg_x.patches:\n",
    "    patch.set_facecolor(color_dca)\n",
    "for patch in fig.ax_marg_y.patches:\n",
    "    patch.set_facecolor(color_ind)\n",
    "\n",
    "plt.subplots_adjust(left=0.2, right=0.8, top=0.8, bottom=0.2) \n",
    "cbar_ax = fig.fig.add_axes([.85, .25, .03, .4])\n",
    "cbar_ax.set_xlabel('\\n Number\\n of sites')\n",
    "plt.colorbar(cax=cbar_ax)\n",
    "    \n",
    "fig.set_axis_labels(\"Context-Dependent\\nEntropy\",\"Context-Independent\\nEntropy\")\n",
    "fig.savefig(Figures_folder / \"Fig_2b.jpeg\", dpi=DPI, bbox_inches=\"tight\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "epsilon = 0.01\n",
    "bins = np.linspace(0-epsilon,np.log2(20)+epsilon,30)\n",
    "\n",
    "with sns.axes_style(\"whitegrid\"):\n",
    "    parameters()\n",
    "    fig = plt.figure(figsize=(WIDTH,WIDTH*HW_RATIO))\n",
    "    ax = fig.add_subplot(111)\n",
    "    sns.histplot(data=stats_df, x=\"CI_Entropy\", kde=True,stat=\"density\",color=color_ind,label=\"Context-Independent Entropy\",ax=ax,bins=bins,line_kws={\"linewidth\":linewidth})\n",
    "    sns.histplot(data=stats_df, x=\"CD_Entropy\", kde=True,stat=\"density\",color=color_dca,label=\"Context-Dependent Entropy\",ax=ax,bins=bins,line_kws={\"linewidth\":linewidth})\n",
    "    ax.set_xlabel(\"Entropy\")\n",
    "    plt.legend()\n",
    "    fig.savefig(Figures_folder / \"Fig_2c.jpeg\", dpi=DPI, bbox_inches=\"tight\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(len(stats_df[stats_df[\"CI_Entropy\"]<1])/len(stats_df))\n",
    "print(len(stats_df[stats_df[\"CD_Entropy\"]<1])/len(stats_df))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Subsection 4"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig = sns.jointplot(x=\"CD_Entropy\",y=\"CI_Entropy\",data=stats_df[stats_df[\"Freq\"]==0],kind=\"hex\",cmap='CMRmap_r',height=WIDTH) #cubehelix_r, gist_earth_r, afmhot_r, hot_r, gnuplot2_r\n",
    "\n",
    "for patch in fig.ax_marg_x.patches:\n",
    "    patch.set_facecolor(color_dca)\n",
    "for patch in fig.ax_marg_y.patches:\n",
    "    patch.set_facecolor(color_ind)\n",
    "    \n",
    "plt.subplots_adjust(left=0.2, right=0.85, top=0.9, bottom=0.3) \n",
    "cbar_ax = fig.fig.add_axes([.9, .35, .03, .5])\n",
    "cbar_ax.set_xlabel('\\n Number\\n of sites')\n",
    "plt.colorbar(cax=cbar_ax)\n",
    "    \n",
    "fig.set_axis_labels(\"Context-Dependent\\nEntropy\",\"Context-Independent\\nEntropy\")\n",
    "fig.fig.suptitle(\"Conserved sites\")\n",
    "fig.savefig(Figures_folder / \"Fig_3a.jpeg\", dpi=DPI, bbox_inches=\"tight\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig = sns.jointplot(x=\"CD_Entropy\",y=\"CI_Entropy\",data=stats_df[stats_df[\"Freq\"]>0.05],kind=\"hex\",cmap='CMRmap_r',height=WIDTH) #cubehelix_r, gist_earth_r, afmhot_r, hot_r, gnuplot2_r\n",
    "\n",
    "for patch in fig.ax_marg_x.patches:\n",
    "    patch.set_facecolor(color_dca)\n",
    "for patch in fig.ax_marg_y.patches:\n",
    "    patch.set_facecolor(color_ind)\n",
    "    \n",
    "plt.subplots_adjust(left=0.2, right=0.85, top=0.9, bottom=0.3) \n",
    "cbar_ax = fig.fig.add_axes([.9, .35, .03, .5])\n",
    "cbar_ax.set_xlabel('\\n Number\\n of sites')\n",
    "plt.colorbar(cax=cbar_ax)\n",
    "    \n",
    "fig.set_axis_labels(\"Context-Dependent\\nEntropy\",\"Context-Independent\\nEntropy\")\n",
    "fig.fig.suptitle(\"Variable sites\")\n",
    "fig.savefig(Figures_folder / \"Fig_3b.jpeg\", dpi=DPI, bbox_inches=\"tight\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "epsilon = 0.01\n",
    "bins = np.linspace(0-epsilon,np.log2(20)+epsilon,30)\n",
    "\n",
    "with sns.axes_style(\"whitegrid\"):\n",
    "    parameters()\n",
    "    fig = plt.figure(figsize=(WIDTH, WIDTH*HW_RATIO))\n",
    "    ax = fig.add_subplot(111)\n",
    "    sns.histplot(data=stats_df[stats_df[\"Freq\"]==0], x=\"CI_Entropy\", kde=True,stat=\"density\",color=color_conserved,label=\"Conserved sites\",ax=ax,bins=bins,line_kws={\"linewidth\":linewidth})\n",
    "    sns.histplot(data=stats_df[stats_df[\"Freq\"]>0.05], x=\"CI_Entropy\", kde=True,stat=\"density\",color=color_variable,label=\"Variable sites\",ax=ax,bins=bins,line_kws={\"linewidth\":linewidth})\n",
    "\n",
    "    ax.set_xlabel(\"Context-Independent Entropy\")\n",
    "    plt.legend()\n",
    "    fig.savefig(Figures_folder / \"Fig_3c.jpeg\", dpi=DPI, bbox_inches=\"tight\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "epsilon = 0.01\n",
    "bins = np.linspace(0-epsilon,np.log2(20)+epsilon,30)\n",
    "\n",
    "with sns.axes_style(\"whitegrid\"):\n",
    "    parameters()\n",
    "    fig = plt.figure(figsize=(WIDTH, WIDTH*HW_RATIO))\n",
    "    ax = fig.add_subplot(111)\n",
    "    sns.histplot(data=stats_df[stats_df[\"Freq\"]==0], x=\"CD_Entropy\", kde=True,stat=\"density\",color=color_conserved,label=\"Conserved sites\",ax=ax,bins=bins,line_kws={\"linewidth\":linewidth})\n",
    "    sns.histplot(data=stats_df[stats_df[\"Freq\"]>0.05], x=\"CD_Entropy\", kde=True,stat=\"density\",color=color_variable,label=\"Variable sites\",ax=ax,bins=bins,line_kws={\"linewidth\":linewidth})\n",
    "\n",
    "    ax.set_xlabel(\"Context-Dependent Entropy\")\n",
    "    plt.legend()\n",
    "    fig.savefig(Figures_folder / \"Fig_3d.jpeg\", dpi=DPI, bbox_inches=\"tight\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "th = 1\n",
    "\n",
    "def entropy(x):\n",
    "    if(x>0):\n",
    "        return -x*np.log2(x)\n",
    "    return 0\n",
    "\n",
    "aa_site_df_1 = aa_site_df\n",
    "aa_site_df_1[\"acc_CI\"] = aa_site_df_1[\"acc_Profile\"].apply(entropy)\n",
    "aa_site_df_1[\"acc_CD\"] = aa_site_df_1[\"acc_DCA_Proba\"].apply(entropy)\n",
    "aa_site_df_1 = aa_site_df_1[[\"Gene\",\"Locus\",\"acc_CI\",\"acc_CD\"]].groupby([\"Gene\",\"Locus\"]).sum().reset_index()\n",
    "df = stats_df[stats_df[\"Freq\"]==0].merge(aa_site_df_1,on=[\"Gene\",\"Locus\"],how=\"inner\")\n",
    "print(len(df[(df[\"acc_CD\"]<th)])/len(df))\n",
    "print(len(df[(df[\"acc_CI\"]<th)])/len(df))\n",
    "df = None\n",
    "\n",
    "print(len(stats_df[(stats_df[\"CI_Entropy\"]<th) & (stats_df[\"Freq\"]==0)])/len(stats_df[(stats_df[\"Freq\"]==0)]))\n",
    "print(len(stats_df[(stats_df[\"CD_Entropy\"]<th) & (stats_df[\"Freq\"]==0)])/len(stats_df[(stats_df[\"Freq\"]==0)]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "epsilon = 0.01\n",
    "bins = np.linspace(0-epsilon,np.log2(9)+epsilon,30)\n",
    "\n",
    "with sns.axes_style(\"whitegrid\"):\n",
    "    parameters()\n",
    "    fig = plt.figure(figsize=(WIDTH, WIDTH*HW_RATIO))\n",
    "    ax = fig.add_subplot(111)\n",
    "    sns.histplot(data=stats_df[stats_df[\"Freq\"]==0].merge(aa_site_df_1,on=[\"Gene\",\"Locus\"]), x=\"acc_CI\", kde=True,stat=\"density\",color=color_conserved,label=\"Conserved sites\",ax=ax,bins=bins,line_kws={\"linewidth\":linewidth})\n",
    "    sns.histplot(data=stats_df[stats_df[\"Freq\"]>0.05].merge(aa_site_df_1,on=[\"Gene\",\"Locus\"]), x=\"acc_CI\", kde=True,stat=\"density\",color=color_variable,label=\"Variable sites\",ax=ax,bins=bins,line_kws={\"linewidth\":linewidth})\n",
    "\n",
    "    ax.set_xlabel(\"Context-Independent Entropy\\n(1-SNP only)\")\n",
    "    plt.legend()\n",
    "    fig.savefig(Figures_folder / \"Fig_S5a.jpeg\", dpi=DPI, bbox_inches=\"tight\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "epsilon = 0.01\n",
    "bins = np.linspace(0-epsilon,np.log2(9)+epsilon,30)\n",
    "\n",
    "with sns.axes_style(\"whitegrid\"):\n",
    "    parameters()\n",
    "    fig = plt.figure(figsize=(WIDTH, WIDTH*HW_RATIO))\n",
    "    ax = fig.add_subplot(111)\n",
    "    sns.histplot(data=stats_df[stats_df[\"Freq\"]==0].merge(aa_site_df_1,on=[\"Gene\",\"Locus\"]), x=\"acc_CD\", kde=True,stat=\"density\",color=color_conserved,label=\"Conserved sites\",ax=ax,bins=bins,line_kws={\"linewidth\":linewidth})\n",
    "    sns.histplot(data=stats_df[stats_df[\"Freq\"]>0.05].merge(aa_site_df_1,on=[\"Gene\",\"Locus\"]), x=\"acc_CD\", kde=True,stat=\"density\",color=color_variable,label=\"Variable sites\",ax=ax,bins=bins,line_kws={\"linewidth\":linewidth})\n",
    "\n",
    "    ax.set_xlabel(\"Context-Dependent Entropy\\n(1-SNP only)\")\n",
    "    plt.legend()\n",
    "    fig.savefig(Figures_folder / \"Fig_S5b.jpeg\", dpi=DPI, bbox_inches=\"tight\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Subsection 5"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "th = 1\n",
    "epsilon = 0.01\n",
    "\n",
    "bins = np.linspace(stats_df[\"IG\"].min()-epsilon,stats_df[\"IG\"].max()+epsilon,40)\n",
    "\n",
    "with sns.axes_style(\"whitegrid\"):\n",
    "    parameters()  \n",
    "    fig = plt.figure(figsize=(WIDTH, WIDTH*HW_RATIO))\n",
    "    ax = fig.add_subplot(111)\n",
    "    sns.histplot(data=stats_df[(stats_df[\"CD_Entropy\"]>=th)], x=\"IG\", kde=True,color=color_variable,label=\"Always variable\",ax=ax,bins=bins,line_kws={\"linewidth\":linewidth})\n",
    "    sns.histplot(data=stats_df[(stats_df[\"CI_Entropy\"]<th)], x=\"IG\", kde=True,color=color_alw_conserved,label=\"Always conserved\",ax=ax,bins=bins,line_kws={\"linewidth\":linewidth})\n",
    "    sns.histplot(data=stats_df[(stats_df[\"CI_Entropy\"]>=th) & (stats_df[\"CD_Entropy\"]<th)], x=\"IG\", kde=True,color=color_cont_conserved,label=\"Context-conserved\",ax=ax,bins=bins,line_kws={\"linewidth\":linewidth})\n",
    "\n",
    "    ax.set_xlabel(\"Information gain (bit)\")\n",
    "    ax.set_ylabel(\"Number of sites\")\n",
    "    ax.set_ylim(0,25000)\n",
    "    ax.set_xlim(-4.35,5.9)\n",
    "    plt.legend(loc='upper right')\n",
    "    fig.savefig(Figures_folder / \"Fig_4b.jpeg\", dpi=DPI, bbox_inches=\"tight\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "th =1\n",
    "print(len(stats_df[(stats_df[\"CI_Entropy\"]<th)])/len(stats_df))\n",
    "print(len(stats_df[(stats_df[\"CI_Entropy\"]>=th) & (stats_df[\"CD_Entropy\"]>=th)])/len(stats_df))\n",
    "print(len(stats_df[(stats_df[\"CI_Entropy\"]>=th) & (stats_df[\"CD_Entropy\"]<th)])/len(stats_df))\n",
    "print(len(stats_df[stats_df[\"IG\"]>1])/len(stats_df))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Subsection 6"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pearson, _ = scipy.stats.pearsonr(double_muts[\"Sum_single_mut\"], double_muts[\"Double_mut\"])\n",
    "slope, intercept, _, _, _ = scipy.stats.linregress(double_muts[\"Sum_single_mut\"], double_muts[\"Double_mut\"])\n",
    "with sns.axes_style(\"whitegrid\"):\n",
    "    parameters() \n",
    "    fig = plt.figure(figsize=(WIDTH,WIDTH*HW_RATIO))\n",
    "    ax = fig.add_subplot(111)\n",
    "    #ax.plot(\"Sum_single_mut\",\"Double_mut\",\".\",data=double_muts, alpha=0.5)\n",
    "    sns.scatterplot(data=double_muts, x=\"Sum_single_mut\", y=\"Double_mut\", ax=ax, ec=\"k\",s=5, alpha=0.1)\n",
    "    #ax.plot([-19.9,23],[-19.9,23],\"--\",color=\"w\",linewidth=1)\n",
    "    ax.set_xlabel(\"∆Ei+∆Ej\")\n",
    "    ax.set_ylabel(\"∆Eij\")\n",
    "    plt.text(3.5, -13, f\"y = {slope:.4f}x + {intercept:.4f}\\n\\nPearson correlation\\ncoefficient: {pearson:.4f}\", bbox=dict(fill=False, edgecolor='k', linewidth=1))\n",
    "    fig.savefig(Figures_folder / \"Fig_5a_modif.jpeg\", dpi=DPI, bbox_inches=\"tight\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(ipr_df[\"Neff\"].median())\n",
    "with sns.axes_style(\"whitegrid\"):\n",
    "    parameters() \n",
    "    fig = plt.figure(figsize=(WIDTH,WIDTH*HW_RATIO))\n",
    "    ax = fig.add_subplot(111)\n",
    "    sns.histplot(data=ipr_df, x=\"Neff\",bins=50)\n",
    "    ax.set_xlabel(\"Effective proportion of residues\\n coupled to an amino acid site\")\n",
    "    ax.set_ylabel(\"Number of sites\")\n",
    "    fig.savefig(Figures_folder / \"Fig_5b.jpeg\", dpi=DPI, bbox_inches=\"tight\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Subsection 7"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "with sns.axes_style(\"whitegrid\"):\n",
    "    parameters()  \n",
    "    fig = plt.figure(figsize=(WIDTH,WIDTH*HW_RATIO))\n",
    "    ax = fig.add_subplot(111)\n",
    "    ax.plot([-0.5,6],[0,0],\"--\",color=\"k\",zorder=0)\n",
    "    sns.boxplot(x=\"Cat\", y=\"Diff\", data=full_seq_df, ax=ax, color=sns.color_palette()[0],linewidth=linewidth, fliersize=linewidth)\n",
    "    ax.set_xlabel(\"Proportion of fixed differences\\nbetween sequences\")\n",
    "    ax.set_ylabel(\"DCA epistatic cost\")\n",
    "    fig.savefig(Figures_folder / \"Fig_6b.jpeg\", dpi=DPI, bbox_inches=\"tight\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "with sns.axes_style(\"whitegrid\"):\n",
    "    parameters() \n",
    "    fig = plt.figure(figsize=(WIDTH,WIDTH*HW_RATIO))\n",
    "    ax = fig.add_subplot(111)\n",
    "    sns.histplot(data=couplings_df, x=\"Coupling\",bins=50)\n",
    "    ax.set_xlabel(\"Epistatic couplings between pairs\\n of fixed differences (E. coli - Y. pestis)\")\n",
    "    ax.set_ylabel(\"Number of pairs\")\n",
    "    ax.set_yscale('log')\n",
    "    fig.savefig(Figures_folder / \"Fig_6c.jpeg\", dpi=DPI, bbox_inches=\"tight\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Simulations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def clean_stats_df(df):\n",
    "    '''\n",
    "    Filter rows in stats file before plotting figures.\n",
    "    '''\n",
    "    ## Filter sites where reference AA is major allele\n",
    "    df = df[df[\"Context_freq\"]>0.5]\n",
    "    ## Filter sites where there's at least one possible synonymous mutation\n",
    "    df = df[(df[\"Total_acc_syn\"]>0)]\n",
    "    ## Compute normalization of observed accessible probability\n",
    "    df[\"Proba_obs_acc_norm\"] = df[\"Proba_obs_acc\"]/df[\"Proba_tot_acc\"]\n",
    "    ## Compute proportion of observed synonymous mutations\n",
    "    df[\"Prop_syn\"] = df[\"Observed_acc_syn\"]/df[\"Total_acc_syn\"]\n",
    "    return df\n",
    "\n",
    "def clean_simulated_sites_df(stats_df,simulated_sites_df):\n",
    "    '''\n",
    "    Keep same genes and loci than stats file, compute probability of reference AA.\n",
    "    '''\n",
    "    simulated_sites_df = stats_df[[\"Gene\",\"Locus\"]].merge(simulated_sites_df,how=\"left\")\n",
    "    simulated_sites_df[\"Ref_proba\"] = simulated_sites_df[\"Proba\"]/simulated_sites_df[\"Acceptance_th\"] - simulated_sites_df[\"Proba\"]\n",
    "    return simulated_sites_df[~simulated_sites_df[\"Proba\"].isna()]\n",
    "\n",
    "stats_df_1 = clean_stats_df(stats_df)\n",
    "simulated_sites = clean_simulated_sites_df(stats_df_1, pd.read_csv(simulated_sites_df_path))\n",
    "\n",
    "lambda_list = [i/10 for i in range(20,51)]\n",
    "nsim = 20\n",
    "M = len(stats_df_1[(stats_df_1[\"Total_acc_syn\"]==4)][\"Prop_syn\"])\n",
    "values = stats_df_1[(stats_df_1[\"Total_acc_syn\"]==4)].groupby(\"Prop_syn\").size().reset_index()[0]\n",
    "\n",
    "## JC69 simulation\n",
    "\n",
    "def syn_simulation(lambda_list, nsim, M):\n",
    "    '''\n",
    "    Simulate synonymous mutations on 4-fold degenerate sites.\n",
    "    '''\n",
    "    probabilities = [1/3]*3\n",
    "    Total = dict()\n",
    "    for lambda_ in lambda_list:\n",
    "        Total[lambda_] = list()\n",
    "        for i in range(nsim):\n",
    "            total = [0 for i in range(4)]\n",
    "            N_list = np.random.poisson(lambda_,M)\n",
    "            for j in range(M):\n",
    "                sample = [1]\n",
    "                values = np.random.choice([i+2 for i in range(len(probabilities))], N_list[j], p=probabilities)\n",
    "                for value in values:\n",
    "                    if((value not in sample) and (random.randint(1, 2)==1)):\n",
    "                        sample.append(value)\n",
    "                total[len(sample)-1] += 1\n",
    "            Total[lambda_].append(total)\n",
    "    return Total\n",
    "\n",
    "def square_loss(Total, values):\n",
    "    '''\n",
    "    Compute and plot square loss to find optimal lambda parameter on simulations of synonymous mutants.\n",
    "    '''\n",
    "    avg = dict()\n",
    "    X = list()\n",
    "    Y = list()\n",
    "    Z = list()\n",
    "    val = np.sum(values*np.array([1,2,3,4]))/np.sum(values)\n",
    "    for key in Total:\n",
    "        avg[key] = list()\n",
    "        for i in Total[key]:\n",
    "            avg[key].append(np.sum(np.square(np.array(i)-values)))\n",
    "        X.append(key)\n",
    "        Y.append(np.mean(np.array(avg[key])))\n",
    "        Z.append(np.std(np.array(avg[key])))\n",
    "    lambda_ = X[np.argmin(np.array(Y))]\n",
    "    mean_values = [list() for i in range(4)]\n",
    "    for i in Total[lambda_]:\n",
    "        for j in range(4):\n",
    "            mean_values[j].append(i[j])\n",
    "    for k in range(4):\n",
    "        mean_values[k] = np.mean(mean_values[k])\n",
    "    return X, Y, Z, mean_values \n",
    "\n",
    "syn_sim_results = square_loss(syn_simulation(lambda_list, nsim, M), values)\n",
    "pd.DataFrame({\"X\":syn_sim_results[0],\"Y\":syn_sim_results[1],\"Z\":syn_sim_results[2]}).to_csv(Data_folder / \"syn_sim_results_1.csv\",index=None)\n",
    "pd.DataFrame({\"Real_values\":values,\"Sim_values\":syn_sim_results[3]}).to_csv(Data_folder / \"synsim_results_2.csv\",index=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "with sns.axes_style(\"whitegrid\"):\n",
    "    parameters()  \n",
    "    X, Y, Z, mean_values = syn_sim_results\n",
    "    fig = plt.figure(figsize=(WIDTH,WIDTH*HW_RATIO))\n",
    "    ax = fig.add_axes([0.1, 0.1, 0.8, 0.8])\n",
    "    ax.errorbar(X, Y, yerr=Z, label='both limits (default)')\n",
    "    ax.set_xlabel(\"λ\")\n",
    "    ax.set_ylabel(\"Square loss\")\n",
    "    fig.savefig(Figures_folder / \"Fig_S3a.jpeg\", dpi=DPI, bbox_inches=\"tight\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.DataFrame({\"Nb_mut\":[1,2,3,4,1,2,3,4],\"Dataset\":[\"Reality\"]*4+[\"Simulation\"]*4,\n",
    "                  \"Nb_site\":list(values)+list(mean_values)})\n",
    "\n",
    "with sns.axes_style(\"whitegrid\"):\n",
    "    parameters()  \n",
    "    fig = sns.catplot(x=\"Nb_mut\", y=\"Nb_site\", hue=\"Dataset\", kind=\"bar\", data=df, height=WIDTH*HW_RATIO, alpha=1, aspect=1/HW_RATIO)\n",
    "    fig.set_axis_labels(\"Number of observed states\",\"Number of sites\")\n",
    "    fig.savefig(Figures_folder / \"Fig_S3b.jpeg\", dpi=DPI, bbox_inches=\"tight\")\n",
    "\n",
    "del df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def aa_simulation(stats,simulated_sites_df,lambda_):\n",
    "    '''\n",
    "    Simulate new (non-synonymous) mutations using the given lambda parameter.\n",
    "    '''\n",
    "    Total_proba = list()\n",
    "    Conserved = list()\n",
    "    for index, row in stats.iterrows():\n",
    "        df = simulated_sites_df[(simulated_sites_df[\"Gene\"]==row['Gene']) & (simulated_sites_df[\"Locus\"]==row['Locus'])]\n",
    "        counts = np.array(list(df[\"Counts\"]))\n",
    "        probabilities = list((counts)/counts.sum())\n",
    "        thresholds = list(df[\"Acceptance_th\"])\n",
    "        AA = list(df[\"AA\"])\n",
    "        ref = list(df[\"Ref\"])[0]\n",
    "        ref_id = -1\n",
    "        if(ref in AA):\n",
    "            ref_id = AA.index(ref)\n",
    "        N = np.random.poisson(lambda_,1)\n",
    "        sample = np.random.choice([i for i in range(len(probabilities))], N, p=probabilities)\n",
    "        accept = [np.random.random(1)[0]<thresholds[i] for i in sample]\n",
    "        accepted_AA = list(np.unique([sample[i] for i in range(len(accept)) if accept[i]]))\n",
    "        if(ref_id in accepted_AA):\n",
    "            accepted_AA.remove(ref_id)\n",
    "        accepted_AA = np.array(accepted_AA)\n",
    "        proba = list(df[\"Proba\"])\n",
    "        Total_proba.append(np.sum([proba[i] for i in accepted_AA])+list(df[\"Ref_proba\"])[0])\n",
    "        Conserved.append(len(accepted_AA)==0)\n",
    "    stats[\"Total_proba\"] = Total_proba\n",
    "    stats[\"Conserved\"] = Conserved\n",
    "    return stats[[\"Gene\",\"Locus\",\"Total_proba\",\"Conserved\"]]\n",
    "\n",
    "X = syn_sim_results[0]\n",
    "Y = syn_sim_results[1]\n",
    "lambda_JC69 = X[np.argmin(np.array(Y))]\n",
    "simulated_aa_jc69 = aa_simulation(stats_df_1[stats_df_1[\"Locus\"]>0],simulated_sites,lambda_JC69)\n",
    "print(lambda_JC69)\n",
    "simulated_aa_jc69.to_csv(\"simulated_aa_jc69.csv\",index=None)\n",
    "stats_df_1[stats_df_1[\"Locus\"]>0].to_csv('stats_1.csv',index=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "with sns.axes_style(\"whitegrid\"):\n",
    "    parameters()  \n",
    "    selected_columns = ['Gene', 'Locus', 'Freq', 'CD_Entropy', 'CI_Entropy', 'Proba_obs_acc', 'Proba_tot_acc']\n",
    "    simulations_df = stats_df_1[stats_df_1[\"Locus\"]>0][selected_columns].merge(simulated_aa_jc69)\n",
    "    fig = plt.figure(figsize=(WIDTH,WIDTH*HW_RATIO))\n",
    "    ax = fig.add_axes([0.1, 0.1, 0.8, 0.8])\n",
    "    ax.hist(simulations_df[simulations_df[\"Freq\"]==0][\"CD_Entropy\"],density=True,bins=[i/5 for i in range(23)], edgecolor='black', linewidth=1, alpha=1, label=\"Reality\")\n",
    "    ax.hist(simulations_df[simulations_df[\"Conserved\"]][\"CD_Entropy\"],density=True,bins=[i/5 for i in range(23)], edgecolor='black', linewidth=1, alpha=0.5, label=\"Simulation\")\n",
    "    plt.legend()\n",
    "    ax.set_xlabel(\"Context-Dependent Entropy\\nof conserved sites\")\n",
    "    ax.set_ylabel(\"Proportion of sites\")\n",
    "    fig.savefig(Figures_folder / \"Fig_S3c.jpeg\", dpi=DPI, bbox_inches=\"tight\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "selected_columns = ['Gene', 'Locus', 'Freq', 'CD_Entropy', 'CI_Entropy', 'Proba_obs_acc', 'Proba_tot_acc']\n",
    "simulations_df = stats_df_1[stats_df_1[\"Locus\"]>0][selected_columns].merge(simulated_aa_jc69)\n",
    "xlim = np.max(simulations_df[\"CD_Entropy\"])+0.1\n",
    "ylim = np.max(simulations_df[\"CI_Entropy\"])+0.1\n",
    "fig = sns.jointplot(x=\"CD_Entropy\",y=\"CI_Entropy\",data=simulations_df[simulations_df[\"Conserved\"]],kind=\"hex\",cmap='CMRmap_r',height=WIDTH)\n",
    "\n",
    "for patch in fig.ax_marg_x.patches:\n",
    "    patch.set_facecolor(color_dca)\n",
    "for patch in fig.ax_marg_y.patches:\n",
    "    patch.set_facecolor(color_ind)\n",
    "\n",
    "plt.subplots_adjust(left=0.2, right=0.8, top=0.8, bottom=0.2) \n",
    "cbar_ax = fig.fig.add_axes([.85, .25, .03, .4])\n",
    "cbar_ax.set_xlabel('\\n Number\\n of sites')\n",
    "plt.colorbar(cax=cbar_ax)\n",
    "\n",
    "fig.set_axis_labels(\"Context-Dependent\\nEntropy\",\"Context-Independent\\nEntropy\")\n",
    "fig.savefig(Figures_folder / \"Fig_S3d.jpeg\", dpi=DPI, bbox_inches=\"tight\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
